client/rest-high-level/src/main/java/org/elasticsearch/client/security/User.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client.security;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+import java.util.List;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+/**
+ * An authenticated user
+ */
+public class User {
+
+    static final ParseField USERNAME = new ParseField("username");
+    static final ParseField ROLES = new ParseField("roles");
+    static final ParseField FULL_NAME = new ParseField("full_name");
+    static final ParseField EMAIL = new ParseField("email");
+    static final ParseField METADATA = new ParseField("metadata");
+    static final ParseField ENABLED = new ParseField("enabled");
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<User, Void> PARSER = new ConstructingObjectParser<>("client_security_user",
+            a -> new User((String) a[0], ((List<String>) a[1]).toArray(new String[0]), (String) a[2], (String) a[3],
+                    (Map<String, Object>) a[4], (Boolean) a[5]));
+    static {
+        PARSER.declareString(constructorArg(), USERNAME);
+        PARSER.declareStringArray(optionalConstructorArg(), ROLES);
```

Comment:
If there would have been a null version for this I would have picked it.
I see the `XContentBuilder` adds `null` for every object type it serializes but the parser is not expecting `null` for each of these, maybe it should?
In addition to the `empty collection` vs `null`, in JSON we could also have the field missing 🎃 Do we have a rule for these?

Reply:
`XContentBuilder` serializes `null` arrays with a `null` JSON value https://github.com/elastic/elasticsearch/blob/b8fb83d7a4663e7ed45a2f42ada826766f956f78/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java#L567 but `ObjectParser` will fail to parse it https://github.com/elastic/elasticsearch/blob/b8fb83d7a4663e7ed45a2f42ada826766f956f78/libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java#L187.